### PR TITLE
Improved OpenCL platform error reports

### PR DESCRIPTION
**Pull request by jenshannoschwalm**
While checking for OpenCL platforms and devices give some better logs and UI hints.

1. As there are timing problems while checking for platforms and devices tell the times in the log
2. more detailed info
3. Report dt preference settings after initializing successfully
4. If there are problems while initializing OpenCL related to platforms or drivers give some plausible reasons via `dt_control_log()`
5. More precise user tips if too many opencl error have been detected
6. some internal opencl functions have been removed from exposing and marked as static

### CODE REVIEW

File: src/develop/pixelpipe_hb.c [hunk 1]
```diff
@@ -2708,8 +2708,12 @@ gboolean dt_dev_pixelpipe_process(
                "[opencl] frequent opencl errors encountered; disabling"
                " opencl for this session!\n");
       dt_control_log
-        (_("darktable discovered problems with your OpenCL setup;"
-           " disabling OpenCL for this session!"));
+        (_("frequent opencl errors encountered; disabling OpenCL for this session!"
```
Comment (TurboGit) [inline src/develop/pixelpipe_hb.c:2711]: I'll remove frequent here, we don't really know if this is frequent in this case and:

`OpenCL errors encountered; disabling it for this session!`

Seems clear enough to me.

File: src/develop/pixelpipe_hb.c [hunk 2]
```diff
@@ -2708,8 +2708,12 @@ gboolean dt_dev_pixelpipe_process(
...
+           " likely reasons:\n"
+           "  opencl out of resources due to preference settings. please try with defaults\n"
```
Comment (TurboGit) [inline src/develop/pixelpipe_hb.c:2713]: Maybe add an hyphen at the start of each line.

File: src/develop/pixelpipe_hb.c (hunk 1 above)
Comment (jenshannoschwalm) [reply src/develop/pixelpipe_hb.c:2711]: right.

File: src/develop/pixelpipe_hb.c (hunk 2 above)
Comment (jenshannoschwalm) [reply src/develop/pixelpipe_hb.c:2713]: yes and in the likely opencl code too.

**Comment by jenshannoschwalm** [comment]
Two questions here:
1. any idea why the CI builds don't work right now?
2. I am quite happy with my recently bought sublime text&merge. Happy to have no pending blanks missing. BUT - i saw you had to correct code style with the lens opencl code. Do you have a tool you are using for this work?

**Comment by TurboGit** [comment]
> 2\. Do you have a tool you are using for this work?

No, doing this manually with Emacs, boring... but to me its important in the long term... so I allocate some time on this.

File: src/develop/pixelpipe_hb.c [hunk 3]
```diff
@@ -2708,8 +2708,12 @@ gboolean dt_dev_pixelpipe_process(
                "[opencl] frequent opencl errors encountered; disabling"
                " opencl for this session!\n");
       dt_control_log
-        (_("darktable discovered problems with your OpenCL setup;"
-           " disabling OpenCL for this session!"));
+        (_("opencl errors encountered; disabling OpenCL for this session!"
```
Comment (TurboGit) [inline src/develop/pixelpipe_hb.c:2711]: opencl -> OpenCL

File: src/develop/pixelpipe_hb.c [hunk 4]
```diff
@@ -2708,8 +2708,12 @@ gboolean dt_dev_pixelpipe_process(
...
+           " likely reasons:\n"
+           "  - opencl out of resources due to preference settings. please try with defaults\n"
```
Comment (TurboGit) [inline src/develop/pixelpipe_hb.c:2713]: OpenCL

**Comment by jenshannoschwalm** [comment]
stupid me with building ...

File: src/common/opencl.c [hunk 5]
```diff
@@ -1152,57 +1154,52 @@ void dt_opencl_init(
...
+                 "[opencl_init] opencl library '%s' found on your system and loaded, preference '%s'\n",
+                 cl->dlocl->library,
+                 (strlen(library) == 0) ? "default path" : library);
   }
 
   all_platforms = malloc(sizeof(cl_platform_id) * DT_OPENCL_MAX_PLATFORMS);
   all_num_devices = malloc(sizeof(cl_uint) * DT_OPENCL_MAX_PLATFORMS);
 
   cl_uint num_platforms = 0;
+  logerror =  "platform detection failed. Some possible reasons:\n"
```
Comment (TurboGit) [inline src/common/opencl.c:1191]: No capital letters :) So Some -> some / Previous -> previous

File: src/develop/pixelpipe_hb.c [hunk 6]
```diff
@@ -2708,8 +2708,12 @@ gboolean dt_dev_pixelpipe_process(
                "[opencl] frequent opencl errors encountered; disabling"
                " opencl for this session!\n");
       dt_control_log
-        (_("darktable discovered problems with your OpenCL setup;"
-           " disabling OpenCL for this session!"));
+        (_("OpenCL errors encountered; disabling OpenCL for this session!"
+           " likely problems:\n"
+           "  - OpenCL out of resources due to preference settings. please try with defaults\n"
```
Comment (TurboGit) [inline src/develop/pixelpipe_hb.c:2713]: add `,` at end of line. Change `.` on the 2 next lines by `,`.

File: src/develop/pixelpipe_hb.c [hunk 7]
```diff
@@ -2708,8 +2708,12 @@ gboolean dt_dev_pixelpipe_process(
...
+           "  - buggy driver for some devive. please run darktable with `-d opencl' to identify.\n"
+           "  - some drivers (for small or old devices) don't support proper number of events.\n"
+           "  - too small headroom settings while using 'use all device memory'"));
```
Comment (TurboGit) [inline src/develop/pixelpipe_hb.c:2716]: And finally add a `.` at end of this line.

File: src/common/opencl.c (hunk 5 above)
Comment (TurboGit) [reply src/common/opencl.c:1191]: For the list of items itself, add a `,` on each line except the last one where you want to add a `.`.

**Review by TurboGit — CHANGES_REQUESTED** [review]
Some clean-up. TIA.

File: src/common/opencl.c (hunk 5 above)
Comment (jenshannoschwalm) [reply src/common/opencl.c]: done

File: src/develop/pixelpipe_hb.c (hunk 7 above)
Comment (jenshannoschwalm) [reply src/develop/pixelpipe_hb.c]: ok again

File: src/develop/pixelpipe_hb.c (hunk 6 above)
Comment (jenshannoschwalm) [reply src/develop/pixelpipe_hb.c]: ok and done

**Review by TurboGit — APPROVED** [review]
Thanks!

File: src/common/opencl.c [hunk 8]
```diff
@@ -1152,57 +1154,52 @@ void dt_opencl_init(
...
-               "[opencl_init] opencl_device_priority: '%s'\n", str);
-  dt_print_nts(DT_DEBUG_OPENCL,
-               "[opencl_init] opencl_mandatory_timeout: %d\n",
-               dt_conf_get_int("opencl_mandatory_timeout"));
 
   // dynamically load opencl runtime and bind required symbols and test it
   cl->dlocl = dt_dlopencl_init(library);
   if(cl->dlocl == NULL)
   {
     logerror = "no working opencl library found";
```
Comment (TurboGit) [inline src/common/opencl.c:1172]: @jenshannoschwalm :
Just missed that, as `logerror` is displayed to user using `dt_control_log()` we need to have those strings translated. So we need `_(` and `)` around all those strings for all `logerror` assignment.

Comment (jenshannoschwalm) [reply src/common/opencl.c:1172]: Yes and sure. Also there have been non-translatable tool tips added for slider buttons. Will fix all in one PR.

**Pull request merged**